pyqtorch/api.py

```diff
@@ -96,12 +99,98 @@ def sample(
     )
 
 
+def analytical_expectation(
+    circuit: QuantumCircuit,
+    state: Tensor,
+    observable: Observable,
+    values: dict[str, Tensor] = dict(),
+    embedding: Embedding | None = None,
+) -> Tensor:
+    """Compute the analytical expectation value.
+
+    Given an initial state :math:`\\ket\\rangle`,
+    a quantum circuit :math:`U(\\theta)`,
+    the analytical expectation value with :math:`O` is defined as
+    :math:`\\langle\\bra U_{\\dag}(\\theta) |O| U(\\theta) \\ket\\rangle`
+
+    Args:
+        circuit (QuantumCircuit): Quantum circuit :math:`U(\\theta)`.
+        state (Tensor): Input state :math:`\\ket\\rangle`.
+        observable (Observable): Observable O.
+        values (dict[str, Tensor], optional): Parameter values for the observable if any.
+        embedding (Embedding | None, optional): An optional instance of `Embedding`.
+
+    Returns:
+        Tensor: Expectation value.
+    """
+    state = run(circuit, state, values, embedding=embedding)
+    return observable.expectation(state, values, embedding=embedding)
+
+
+def sampled_expectation(
+    circuit: QuantumCircuit,
+    state: Tensor,
+    observable: Observable,
+    values: dict[str, Tensor] = dict(),
+    embedding: Embedding | None = None,
+    n_shots: int = 1,
+) -> Tensor:
+    """Expectation value approximated via sampling.
+
+    Given an input state :math:`\\ket\\rangle`,
+    the expectation analytical value with :math:`O` is defined as
+    :math:`\\langle\\bra|O\\ket\\rangle`
+
+    Args:
+        state (Tensor): Input state :math:`\\ket\\rangle`.
+        observable (Observable): Observable O.
+        values (dict[str, Tensor], optional): Parameter values for the observable if any.
+        embedding (Embedding | None, optional): An optional instance of `Embedding`.
+        n_shots: (int, optional): Number of samples to compute expectation on.
+
+    Returns:
+        Tensor: Expectation value.
+    """
+    state = run(circuit, state, values, embedding=embedding)
+    n_qubits = circuit.n_qubits
+
+    # batchsize needs to be first dim for eigh
+    eigvals, eigvecs = torch.linalg.eigh(
+        observable.tensor(values=values, embedding=embedding, full_support=list(range(n_qubits))).permute((2, 0, 1))
+    )
+    eigvals = eigvals.squeeze()
+    eigvec_state_prod = apply_operator(
+        state,
+        eigvecs.T.conj(),
+        tuple(range(n_qubits)),
+        n_qubits=circuit.n_qubits,
+    )
```

Comment:
You can make this more efficient by doing `observable.tensor()` without the `full_support`, calling `eigh`, and then expanding the `eigenvalues` and `eigenvectors` if needed. See the comment I made [here](https://github.com/pasqal-io/horqrux/pull/23#discussion_r1692698274) on horqrux (actually just today xD).

Reply:
After you get the `eigvecs` matrix it might be enough to call `expand_operator` on it, but best to double-check this.

Reply:
@jpmoutinho I think I will tackle this in another MR. This requires a subtle logic for the eigenvalues as well.